### PR TITLE
Check GMT_DOC_DIR if SHAREDIR fails

### PR DESCRIPTION
**Pull request by PaulWessel**
While we configured GMT_DOC_DIR it was never used.  Now it is, and hopefully this PR addresses #1960.

### CODE REVIEW

**Comment by seisman** [comment]
Doesn't work. See the line 2 below.

```
docs [DEBUG]: Try URL path: file:////usr/local/Cellar/gmt/6.0.1/share/gmt/doc/html/coast.html
docs [DEBUG]: No access, Now try URL path: file:////usr/local/Cellar/gmt/6.0.1//usr/local/Cellar/gmt/6.0.1/share/doc/gmt/html/coast.html
docs [DEBUG]: No local access,use server URL path: https://docs.generic-mapping-tools.org/6.0/coast.html
docs [DEBUG]: Opening https://docs.generic-mapping-tools.org/6.0/coast.html via open
gmt [DEBUG]: Entering GMT_Destroy_Session
```

**Comment by PaulWessel** [comment]
So line two:

`file:////usr/local/Cellar/gmt/6.0.1//usr/local/Cellar/gmt/6.0.1/share/doc/gmt/html/coast.html
`

 is the result of

`snprintf (URL, PATH_MAX, "file:///%s/html/%s", GMT_DOC_DIR, module);
`

Thus **GMT_DOC_DIR** when brew configures things is 

`/usr/local/Cellar/gmt/6.0.1//usr/local/Cellar/gmt/6.0.1/share/doc/gmt
`

which is not too surprising given the config.h.in statement

```
/* path to documentation */
#define GMT_DOC_DIR "@CMAKE_INSTALL_PREFIX@/@GMT_DOCDIR@"
```

It sounds like this one simply must be 

`#define GMT_DOC_DIR "@GMT_DOCDIR@"
`

**Comment by seisman** [comment]
We probably need to check if GMT_DOCDIR is given as an absolute path (like in homebrew) or a relative path (like the GMT default path).

**Comment by PaulWessel** [comment]
Yes, so when I just build in a local build dir parallel to src I have it install in build/gmt6/{bin, lib,..} so then GMT_DOC_DIR is "share/doc" and the @CMAKE_INSTALL_PREFIX@/@GMT_DOCDIR@ is what I need.

As for config.h it has

```
#define GMT_SHARE_DIR "/Users/pwessel/UH/RESEARCH/CVSPROJECTS/GMTdev/gmt-dev/dbuild/gmt6/share"
#define GMT_DOC_DIR "share/doc"
```

so maybe try both GMT_DOC_DIR and GMT_SHARE_DIR?

File: src/config.h.in [hunk 1]
```diff
@@ -51,7 +51,7 @@
 #define GMT_BINARY_DIR_SRC_DEBUG "@GMT_BINARY_DIR@/src"
 
 /* path to documentation */
-#define GMT_DOC_DIR "@CMAKE_INSTALL_PREFIX@/@GMT_DOCDIR@"
+#define GMT_DOC_DIR "@GMT_DOCDIR@"
```
Comment (seisman) [inline src/config.h.in]: I'm a little confused here. Does it mean `@GMT_DOCDIR@` must be an absolute path?

**Comment by seisman** [comment]
In `cmake/ConfigUserTemplate.cmake`, we claim that all the paths (GMT_BINDIR, GMT_LIBDIR, GMT_INCLUDEDIR, GMT_DATADIR, GMT_DOCDIR, GMT_MANDIR) can be either absolute or relative paths.
https://github.com/GenericMappingTools/gmt/blob/816825dbaf959a9a6679033e04cecccd8bd32a36/cmake/ConfigUserTemplate.cmake#L67-L71

However, it seems we assume that these paths must be relative paths, e.g.
https://github.com/GenericMappingTools/gmt/blob/816825dbaf959a9a6679033e04cecccd8bd32a36/src/config.h.in#L36-L42

File: src/config.h.in (hunk 1 above)
Comment (PaulWessel) [reply src/config.h.in]: Well, the way it was (line two above) you got two hard paths added together.

Comment (seisman) [reply src/config.h.in]: Yes, but if someone gives GMT_DOCDIR a relative path, then `GMT_DOC_DIR` is also a relative path.

Comment (PaulWessel) [reply src/config.h.in]: Well, I can put it back to what it was but then we get line 2 for brew.
We could add a GMT_INTALLDIR="@CMAKE_INSTALL_PREFIX@" to config.h to have a separate parameter for that, but I guess I don't understand exactly how the package managers are specifying things like doc dir, i.e., relative (to what?) or absolute.

Comment (seisman) [reply src/config.h.in]: Homebrew uses `-DGMT_DOCDIR=#{share}/doc/gmt`. I don't know exactly what `#{share}` means in homebrew, but probably homebrew should use `-DGMT_DOCDIR=share/doc/gmt`, which is relative to `${CMAKE_INSTALL_PREFIX}`.

**Comment by PaulWessel** [comment]
Would the brew guy know?

**Comment by seisman** [comment]
https://docs.brew.sh/Formula-Cookbook#variables-for-directory-locations

It seems `#{share}` is `/usr/local/Cellar/gmt/6.0.0/share`. Homebrew definitely should use a relative path, and we should claim that all the paths (e.g. GMT_DOCDIR) must be relative paths.

**Comment by PaulWessel** [comment]
OK with me.  So leave my changes to config.h.in then (removing the INSTALL PREFIX)?

**Comment by seisman** [comment]
The change to config.h.in should be reverted.

**Comment by PaulWessel** [comment]
You mean it should be what it was, i.e.,

`#define GMT_DOC_DIR "@CMAKE_INSTALL_PREFIX@/@GMT_DOCDIR@"
`

**Comment by seisman** [comment]
Yes, all the path settings in `cmake/ConfigUserTemplate.cmake` should be relative to `CMAKE_INSTALL_PREFIX`.

**Comment by seisman** [comment]
@claudiodsf I made some changes to the homebrew recipe (attached here [gmt.rb.zip](https://github.com/GenericMappingTools/gmt/files/3801140/gmt.rb.zip)).

1. Apply two patches in commit https://github.com/GenericMappingTools/gmt/commit/b65dc6ebe7eba396b57dbad59e85b7ba7e8e2908 and [the current PR](https://patch-diff.githubusercontent.com/raw/GenericMappingTools/gmt/pull/1964.diff)
2. Change `GMT_DOCDIR` and `GMT_MANDIR` to relative paths.

Then I reinstalled GMT with `brew install --build-from-source -vv ./gmt.rb`. It now works for me. It would be better if you can also give it a try.

**Comment by claudiodsf** [comment]
I confirm that it now works and that the new formula has the same installation tree than the previous one.
I'll add a note to the formula for future maintainers on that `GMT_DOCDIR` and `GMT_MANDIR` must be relative paths.
I think we can keep this for 6.0.1, which should be not so far in the future, right?

**Comment by seisman** [comment]
@claudiodsf Not sure when we'll release 6.0.1. Since "gmt docs" is the main new feature in GMT6, and many macOS users use homebrew, I suggest applying the two patches I mentioned before and bumping the homebrew revision number (i.e. 6.0.0-1).

**Comment by claudiodsf** [comment]
Ok, I submitted the PR with version bump to homebrew-core: https://github.com/Homebrew/homebrew-core/pull/46231

**Pull request merged**